### DISCUSSION TITLE
[창훈] assignment 기능구현

### DISCUSSION THREAD

**Discussion by anotheranotherhoon** [Wanted-07-team-9/wanted_assignment_02#5] [Ideas]
### 이슈 목록 가져오기 API 활용

```js
export const getIssue = async (pageNumber) => {
  const { data } = await axiosInstance.get(`/repos/angular/angular-cli/issues?sort=comments&page=${pageNumber}&state=open`)
      return data
}
```

- [x] open 상태의 이슈 중 코멘트가 많은 순으로 정렬
- [x] 각 행에는 '이슈 번호, 이슈제목, 작성자, 작성일, 코멘트 수' 표시
- [x] 다섯번째 셀에는 광고 이미지 출력

### 화면을 아래로 스크롤 할 시 이슈 목록 추가 로딩 (infinite scrull)

```js
const initialState = {
  loading: false,
  data : [],
  pageNumber : 1,
  error : null,
  reachedEnd: false
}

  const fetchIssues = async(initialState) => {
    dispatch({type:'LOADING'})
    try{
      const response = await getIssue(initialState)
      if(response.length>0){
        dispatch({type:'SUCCESS', data: response})
      }else{
        dispatch({type:'REACH_END', data: response})
      }
    } catch(error) {
      dispatch({type:'ERROR', error})
  }
  }

useEffect(()=>{
    fetchIssues(initialState.pageNumber)
  }, [])
```

```js
 case "SUCCESS" : {
      return {
        loading: false,
        data :[...state.data, ...action.data],
        pageNumber: state.pageNumber +1,
        error : null,
        reachedEnd: false
      }
    }
    case "REACH_END":{
      return {
        loading: false,
        data : state.data,
        pageNumber: state.pageNumber,
        error : null,
        reachedEnd: true
      }
    }
```
- [ ] 위 코드와 같이 무한 스크롤을 구현하려 하였으나 useEffect와 pageNumber 가 마음대로 작동하지 않아 다른 방법을 참고하여 구현함
- [x]Observer API 를 이용하여 무한 스크롤을 구현하였습니다. 

## Assignment2. Context API를 활용한 API 연동
![스크린샷 2022-10-31 오전 3 47 43](https://user-images.githubusercontent.com/77476348/198896125-34ba4c56-572c-4acc-8be9-b7e3fd2747e6.png)
### 이슈의 상세 내용 표시
- [x] '이슈 번호, 이슈 제목, 작성자, 코멘트 수, 작성자 프로필 이미지, 본문' 표시

## Assignment3. 두 페이지 공통 헤더 공유
- [x] 두 페이지는 공통 헤더를 공유
- [x] 헤더에는  Organization Name / Repository Name가 표시
```js
const Container = ({children}) => {
  const title = 'Angular / Angular-cli'
  return(
    <Layout>
          <Header>{title}</Header>
          {children}
    </Layout>

  )
}
```
두 페이지 모두 Container를 공유하게 하였습니다.

## 필수 요구사항
- [x] 이슈 목록 및 상세 화면 기능 구현 
- [x] Context API를 활용한 API 연동 
- [x] 데이터 요청 중 로딩 표시

```js
const Signal = ({ status, imgSrc }) => {
  return (
    <SignalWrapper>
      <IMG src={imgSrc} alt='에러' />
      {status}
    </SignalWrapper>
  )
}

{isLoading && <Signal status={'로딩 중'} imgSrc={LoadingIndicator} />}
```
isLoading 가 참 일 경우 Signal 컴포넌트에 props로 메세지와 이미지를 내려보냄
에러, 로딩 상태에서 Signal컴포넌트를 공유함

## Assignment4. 에러 화면 구현

![스크린샷 2022-10-31 오전 3 48 17](https://user-images.githubusercontent.com/77476348/198896149-4e6dce6a-bb76-4a1d-a298-24ea636fd08f.png)

## Assignment5. 지정된 조건(open 상태, 코멘트 많은 순)에 맞게 데이터 요청 및 표시
- [x] 지정된 조건(open 상태, 코멘트 많은 순)에 맞게 데이터 요청 및 표시

## Assignment6. 반응형 웹 구현(UI는 데스크톱, 모바일에서 보았을 때 모두 읽기 편하게 구현)
- 반응형 웹 구현
```js
  body {
    font-family: 'Oswald', sans-serif;
    font-size: 1.5rem;
    @media screen and (max-width: 413px) {
      font-size: 1rem;
    }
  }
```
paddig, margin을 줄 때 고정 값이 아닌 %를 사용하였습니다!

**Comment by songseul** [comment]
# 화면을 아래로 스크롤 할 시 이슈 목록 추가 로딩 (infinite scroll)
 - 리듀서 방식으로 무한 스크롤을 구현을 시도 했다는 점
 - 코드의 가독성이 높아 졌고 무한 스크롤 함수를 추상화 하여 사용할 수 있다는 것이 좋은것 같습니다.

**Comment by jun-05** [comment]
- 저도 리듀서를 사용해서 무한 스크롤을 구현하신게 대단하신거 같아요.

**Comment by tirhande** [comment]
헤더 공유 부분을 하나의 컴포넌트(Container.js)로 묶으신 부분이 좋았어요!
